File: LeetCode/149.Max_Points_on_a_Line/149.Max_Points_on_a_Line.py
#!/usr/bin/env python3


class Solution:
    def maxPoints(self, points: List[List[int]]) -> int:
        n = len(points)
        res = 0
        for i, p in enumerate(points):
            counter = defaultdict(int)
            xi, yi = p
            overlap = 0
            max_points = 0

            for j in range(n):
                xj, yj = points[j]
                if i == j:
                    continue
                if xi == xj and yi == yj:
                    overlap += 1
                    continue
                dx, dy = xi - xj, yi - yj  # 4, 6

                g = gcd(dx, dy)  # 2
                slope = (dy // g, dx // g)  # 2,3
                counter[slope] += 1
                max_points = max(max_points, counter[slope])

            res = max(res, max_points + 1 + overlap)

        return res

File: LeetCode/149.Max_Points_on_a_Line/149.Max_Points_on_a_Line.java
class Solution {
    public int maxPoints(int[][] points) {
        int n = points.length;
        if (n < 3)
            return n;

        int max = 2;
        for (int i = 0; i < n; i++) {
            Map<Double, Integer> map = new HashMap<>();
            int localMax = 1;
            for (int j = 0; j < n; j++) {
                if (i == j)
                    continue;
                int deltaX = points[i][0] - points[j][0];
                int deltaY = points[i][1] - points[j][1];

                Double slope = (deltaY == 0) ? Double.POSITIVE_INFINITY : (double) deltaX / deltaY;
                map.put(slope, map.getOrDefault(slope, 1) + 1);
                localMax = Math.max(localMax, map.get(slope));
            }
            max = Math.max(max, localMax);
        }
        return max;
    }
}